File: leetcode/lt1.go
package main

import (
	"fmt"
	"sort"
)

type numAndIndex struct {
	num   int
	index int
}

func twoSum(nums []int, target int) []int {
	src := make([]numAndIndex, 0, len(nums))
	for idx, num := range nums {
		src = append(src, numAndIndex{num, idx})
	}
	sort.Slice(src, func(i, j int) bool {
		return src[i].num < src[j].num
	})
	idx1 := 0
	idx2 := len(src) - 1

	for idx1 < idx2 {
		sum := src[idx1].num + src[idx2].num
		if sum == target {
			return []int{src[idx1].index, src[idx2].index}

		} else if sum < target {
			idx1++
		} else {
			idx2--
		}
	}
	return nil
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9
	ans := twoSum(nums, target)
	fmt.Println(ans)
}

File: leetcode/lt1.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> twoSum(vector<int>& nums, int target) {
    map.clear();
    vector<int> ret;
    for (int i = 0; i < nums.size(); ++i) {
      int t_num = target - nums[i];
      if (map.find(t_num) != map.end()) {
        ret.push_back(i);
        ret.push_back(map[t_num]);
      } else {
        map[nums[i]] = i;
      }
    }
    return ret;
  }

  std::unordered_map<int, int> map;
};

int main() {
  Solution s;
  {
    std::vector<int> nums{2, 7, 11, 5};
    int target = 9;
    std::vector<int> ans = s.twoSum(nums, target);
    printf("%d %d\n", ans[0], ans[1]);
  }

  return 0;
}